### PR TITLE
Backup and restore stubbed method to and from prepended module

### PR DESCRIPTION
**Pull request by texpert**
Define the stub method directly on the last prepended method on which the target method has been defined..

These changes helped to fix the test suite of a project I'm working on, where I've started using prepended modules. Ruby 2.3.7 and Rails 4.1.16

### CODE REVIEW

File: lib/rspec/mocks/any_instance/recorder.rb [hunk 1]
```diff
@@ -217,7 +220,7 @@ def restore_original_method!(method_name)
             # @klass originally owned.
             #
             # https://github.com/ruby/ruby/commit/c8854d2ca4be9ee6946e6d17b0e17d9ef130ee81
-            if RUBY_VERSION < "2.3" || backed_up_method_owner[method_name.to_sym] == self
+            if RUBY_VERSION < "2.3" || method_owner == self
```
Comment (godfat) [inline lib/rspec/mocks/any_instance/recorder.rb]: I do not really understand what this was trying to do, but the patch you're providing is simply making this always equals to true, from what I can see, because `self` is in the context of `method_owner`, therefore `method_owner` of course would be `self`.

The original code was comparing `method_owner` with `@klass`, which is a different case.

File: lib/rspec/mocks/any_instance/recorder.rb [hunk 2]
```diff
@@ -234,9 +237,11 @@ def backup_method!(method_name)
           return unless public_protected_or_private_method_defined?(method_name)
 
           alias_method_name = build_alias_method_name(method_name)
-          @backed_up_method_owner[method_name.to_sym] ||= @klass.instance_method(method_name).owner
-          @klass.class_exec do
+          method_owner = @klass.instance_method(method_name).owner
+          @backed_up_method_owner[method_name.to_sym] ||= method_owner
+          method_owner.class_exec do
```
Comment (godfat) [inline lib/rspec/mocks/any_instance/recorder.rb]: I think this can be wrong too. The method owner might be from a module, but we won't want to touch the module, but the `@klass`

**Review by godfat — COMMENTED** [review]
I have some troubles running the test suite so I didn't try to verify this yet, but would you be able to put a test here which can prove that this will fix something? Beside that I have some questions for you. I commented inline.

And just to be clear, this is not the official rspec-mocks repository, so even if I merge it it's unlikely to go into the official branch soon.

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 1 above)
Comment (texpert) [reply lib/rspec/mocks/any_instance/recorder.rb]: Yeah, I think this check could be removed at all.

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 2 above)
Comment (texpert) [reply lib/rspec/mocks/any_instance/recorder.rb]: I have debugged this thoroughly. Prior to this PR:

- the method was aliased on the @klass, but not removing the original method, so stubbing wasn't working - the original method stayed in place
- trying to add `remove_method method_name` after the `alias_method alias_method_name, method_name` brought the `method `sync_product_to_flow' not defined in #<Spree::Variant not initialized>`

That's why I have decided that it is better to alias and remove the original method on its owner. The stub is created directly on the @klass, and removed from the @klass too, while restoring the original method on the owner.

**Comment by texpert** [comment]
> I have some troubles running the test suite so I didn't try to verify this yet, but would you be able to put a test here which can prove that this will fix something? 

I have not managed to set properly the test suite. Have tried to change the gemspec and Gemfile locally with some hardcoded versions for rspec gems, but got a suite, that was giving spurious test failures - each time different - from 8 to a few hundreds of failures :(

> And just to be clear, this is not the official rspec-mocks repository, so even if I merge it it's unlikely to go into the official branch soon.

That's not a problem for my project - I have forked you repo with this branch, implemented my changes, and our test suite is doing well.

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 2 above)
Comment (godfat) [reply lib/rspec/mocks/any_instance/recorder.rb]: I'll take a closer look tomorrow, but want to let you know the case I have in mind. What would happen here?

``` ruby
module M
  def f
    'f'
  end
end

class C0
  include M
end

class C1
  include M
end

c0 = C0.new
c1 = C1.new

allow(c0).to receive(:f).and_return('g')

c0.f # => g
c1.f # => f?
```

Comment (texpert) [reply lib/rspec/mocks/any_instance/recorder.rb]: Yeah, the method will be changed for both classes, for sure. But then, to make it work for just one class, there should be some more work done on stubbing - like the new method analyzing the context and switching to original method if it finds itself in the non-target class.

Comment (godfat) [reply lib/rspec/mocks/any_instance/recorder.rb]: I just tested. The current behaviour will give `f` for `c1.f`, so I don't think this is acceptable as it's changing the behaviour, and I don't think it should change for both classes :(

If we're stubbing on `C0` not on method owner, it won't affect `C1`. So I think this is the wrong fix. I suggest you put some simple tests to show what we're trying to fix here. Otherwise I don't know what I can suggest beside pointing out that this can change the behaviour which is not desirable.

Comment (texpert) [reply lib/rspec/mocks/any_instance/recorder.rb]: Oh, my bad, I thought of `allow_any_instance_of` instead of `allow`.

It is working OK with instance stubbing.
Screenshot 2020-12-04 at 20 12 17

Comment (texpert) [reply lib/rspec/mocks/any_instance/recorder.rb]: But with `allow_any_instance_of `, we got `c1.f undefined method `f' for #<C1:0x00007fcd8b06f350>`

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 1 above)
Comment (godfat) [reply lib/rspec/mocks/any_instance/recorder.rb]: Well, I think that's because you changed the aliased method to be on the method owner, therefore it must be coming from the method owner. However, originally it's from the `@klass` which might or might not be the method owner.

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 2 above)
Comment (godfat) [reply lib/rspec/mocks/any_instance/recorder.rb]: > the method was aliased on the @Klass, but not removing the original method, so stubbing wasn't working - the original method stayed in place

I am confused. Why should it remove the original method? Can you show a specific example?

>  But with allow_any_instance_of , we got c1.f undefined method f' for #C1:0x00007fcd8b06f350`

I tried `allow_any_instance_of` on both `C1`, `C2`, and `M`. Without your patch it's all working as expected to me. So I am still not clear what you're trying to fix, and I don't see the error you're mentioning without your patch.

**Review by godfat — COMMENTED** [review]
I'll try to rebase at some point, and it might fix the test suite. However I would like to know what you're trying to fix first.

**Comment by texpert** [comment]
> I'll try to rebase at some point, and it might fix the test suite. However I would like to know what you're trying to fix first.

I am trying to fix working of `allow_any_instance_of` when prepending the modules (when including the modules it works OK, yes):

```
require 'rails_helper'

class C0
end

module M
  def f
    'f'
  end

  ::C0.prepend(self) if ::C0.included_modules.exclude?(self)
end

class C1
  include M
end

c0 = C0.new
c1 = C1.new

RSpec.describe C0 do
  it do
    # allow(c0).to receive(:f).and_return('g')
    allow_any_instance_of(C0).to receive(:f).and_return('g')

    c0.f # => g
    c1.f # => f?
  end
end
```

Without the patch is failing:

```
0) C0 
     Failure/Error: allow_any_instance_of(C0).to receive(:f).and_return('g')
       Using `any_instance` to stub a method (f) that has been defined on a prepended module (M) is not supported.
     # ./spec/services/prepend_test_spec.rb:24:in `block (2 levels) in <top (required)>'
```

With the patch the `allow_any_instance_of` is working OK, but we got c1.f undefined method f' for #C1:0x00007fcd8b06f350`.

**Comment by godfat** [comment]
Thank you for the simple test which I can verify. I understand what you're trying to fix now, but I believe it's the wrong fix. I tried the same test on my mocking library, [Muack](https://github.com/godfat/muack), and it's not working as expected either.

Both `c0.f` and `c1.f` would give `'g'`. 

I think I understand why it's happening. The problem here is that the prepended module comes first than the singleton class, therefore it's acting on the prepended module `M` here, rather than the singleton class of `C0`.

In order to fix this, we'll need to add something before the prepended module, which, could be another generated prepended module. This is sad because there's no way to remove the prepended module therefore it means the class cannot be reset back to the original state.

But maybe we can minimize the damage by making it a truly transparent module? I have no answer yet. I'll see if I can fix this in my mocking library first, and see if that knowledge can be applied back to rspec-mocks because I am much more familiar with my mocking library, making it much easier to experiment potential solution.

**Comment by godfat** [comment]
So I committed https://github.com/godfat/muack/commit/dff0842139b06ac0dcc72599a05f84b421791a9b which should fix most of the use cases. It is essentially doing exactly what I described above: Prepend an additional module whenever we detected that there's a module being prepended and on the way.

``` ruby
def self.prepare_target singleton_class
  target = singleton_class.ancestors.first
  if target == singleton_class # no prepended module detected
    target
  else # prepend an internal module to override the prepended module(s)
    name = :MuackPrepended
    if singleton_class.const_defined?(name)
      singleton_class.const_get(name)
    else
      prepended = ::Module.new
      singleton_class.const_set(name, prepended)
      singleton_class.private_constant(name)
      singleton_class.prepend(prepended)
      prepended
    end
  end
end
```

And we mock on the newly prepended module to "override" the prepended module. Maybe this can be more precise that this is only needed when the method name was conflicted. I'll see if I can improve this part and see how that would be different.

I think the same idea can be applied to rspec-mocks but I am not fully confident where it should be applied and if this can be accepted. I will try to get there at some point, but no promise though, since I don't really want to maintain rspec-mocks.

**Comment by texpert** [comment]
Great!

I have been exactly thinking of dynamically prepending a Module, as per your suggestion 7 hours ago.

Maybe even not need to analyze is the method from a prepended Module, or not, just always prepend the stub?

**Comment by texpert** [comment]
@godfat , could you, please, rebase the https://github.com/rspec/rspec-mocks/pull/1218 on a fresh `main`so maybe to fix the test suite?

**Comment by godfat** [comment]
> Maybe even not need to analyze is the method from a prepended Module, or not, just always prepend the stub?

@texpert I think it would be much better to only prepend a module when it's absolutely needed, because there's no way we can remove this newly prepended module and that's changing the application code for the lifetime of the process, which can be breaking application codes which might be relying on ancestors chain being intact.

For example, this tool will be quite sensitive to that: https://gitlab.com/gitlab-org/gitlab/-/blob/master/lib/gitlab/utils/override.rb

I just pushed another commit to make this conditional: https://github.com/godfat/muack/commit/e8bd6473ae58b805cbc0101fb2bc14eff9c379fc The critical check is:

``` ruby
index = singleton_class.ancestors.index(singleton_class)
prepended_modules = singleton_class.ancestors[0...index]

if prepended_modules.find{ |m| direct_method_defined?(m, msg) }
  # .. an internal prepend is needed
else
  # can just use the singleton class
end
```

There's only one case which I think it's probably impossible to solve right now:

``` ruby
m0 = Module.new{ def f; :m0; end }
m1 = Module.new{ def f; :m1; end; prepend m0 }
c0 = Class.new{ prepend m0 }.new
c1 = Class.new{ prepend m1 }.new

mock(any_instance_of(m1)).f{:g}

expect(c0.f).eq :m0
expect(c1.f).eq :g # m1 does not know c1 thus no way to pass this
```

I am skipping this test for now, but I hope in the future it's possible to solve this.

> could you, please, rebase the rspec#1218 on a fresh mainso maybe to fix the test suite?

Yes, I have rebased it, and it's great to see that all tests just passed now, and I didn't even change anything.

**Comment by texpert** [comment]
I've managed to implement stubbing the last prepended module, so no need to create dynamically another one prepended :)

**Comment by godfat** [comment]
> I've managed to implement stubbing the last prepended module, so no need to create dynamically another one prepended :)

What about this case?

``` ruby
module M
  def f
    :f 
  end
end

class C0
  prepend M
end

class C1
  prepend M
end

RSpec.describe do
  it do
    allow_any_instance_of(C0).to receive(:f).and_return(:g)

    expect(C0.new.f).to eq(:g)
    expect(C1.new.f).to eq(:f)
  end
end
```

If you're stubbing the last prepended module, I would guess now stubbing on `C0` would affect `C1`, which is not desired.

That's where why we need to prepend a new module, because the prepended module can be used by the completely unrelated class.

**Comment by texpert** [comment]
Seems that I've managed to do it, finally, for all the cases - added few specs for these cases too.

**Comment by godfat** [comment]
Hopefully I'll review tomorrow.

File: lib/rspec/mocks/any_instance/recorder.rb [hunk 3]
```diff
@@ -123,6 +124,14 @@ def build_alias_method_name(method_name)
           "__#{method_name}_without_any_instance__"
         end
 
+        # @private
+        def build_alias_prepended_method_name(method_name)
+          unless @klass.name
+            Object.const_set("Klass_#{@klass.object_id}", @klass)
```
Comment (godfat) [inline lib/rspec/mocks/any_instance/recorder.rb:130]: Do we really need this? Setting a constant at a global namespace feels wrong to me, and we should somehow clean this up in the cleanup process.

File: lib/rspec/mocks/any_instance/recorder.rb [hunk 4]
```diff
@@ -123,6 +124,14 @@ def build_alias_method_name(method_name)
...
+          end
+          "__#{@klass.name}_#{method_name}_without_any_instance__"
```
Comment (godfat) [inline lib/rspec/mocks/any_instance/recorder.rb:132]: We can probably just use `"Klass_#{@klass.object_id}"` as the class name here because they're the same anyway, and we avoid touching the global namespace.

``` ruby
"__Klass_#{@klass.object_id}_#{method_name}_without_any_instance__"
```

File: spec/rspec/mocks/any_instance_spec.rb [hunk 5]
```diff
@@ -176,12 +180,44 @@ def private_method; :private_method_return_value; end
...
+          it 'allows stubbing a method that is defined on the prepended module' do
             klass.class_eval { prepend Module.new { def foo; end } }
+            allow_any_instance_of(klass).to receive(:foo).and_return(45)
 
-            expect {
+            expect(klass.new.foo).to eq(45)
+          end
+
+          context 'different classes having the same module prepended' do
+            let(:klass1) do
```
Comment (godfat) [inline spec/rspec/mocks/any_instance_spec.rb]: Should probably be called something else, because `klass1` is a bit confusing. Maybe `other_klass`.

File: lib/rspec/mocks/any_instance/recorder.rb [hunk 6]
```diff
@@ -275,18 +328,9 @@ def mark_invoked!(method_name)
...
-            return unless problem_mod
-
-            AnyInstance.error_generator.raise_not_supported_with_prepend_error(method_name, problem_mod)
-          end
-        else
-          def allow_no_prepended_module_definition_of(_method_name)
-            # nothing to do; prepends aren't supported on this version of ruby
-          end
+        def prepended_module(method_name)
+          @prepended_module =
```
Comment (godfat) [inline lib/rspec/mocks/any_instance/recorder.rb:345]: I can't tell if this might be causing a problem for now, but I think this is too stateful and could easily lead to bugs. Consider this case:

``` ruby
prepended_module(:test0)
@prepended_module # this is based on :test0
prepended_module(:test1)
@prepended_module # this is based on :test1
```

This memorization should be based on `method_name`, or we just don't memorize at all.

File: lib/rspec/mocks/any_instance/recorder.rb [hunk 7]
```diff
@@ -275,18 +328,9 @@ def mark_invoked!(method_name)
...
+            RSpec::Mocks::Proxy.prepended_modules_of(@klass).select { |mod| mod.method_defined?(method_name) }.last
```
Comment (godfat) [inline lib/rspec/mocks/any_instance/recorder.rb]: Three things here.

0. `method_defined?` does not cover private methods, so `private_method_defined?` should also be used.
1. It should not look for super methods, therefore `false` should be passed to `method_defined?` too. Unfortunately, older Rubies do not accept more arguments, we'll need to fall back to `instance_methods(false).include?(method_name)`
2. Why is it on `last` prepended module? What would happen if there are multiple prepended modules which all have the same method?

For the second point, consider this case:

```
>> Class.new.method_defined? :to_s
=> true
>> Class.new.method_defined? :to_s, false
=> false
```

How I implemented it in Muack: https://github.com/godfat/muack/blob/e8bd6473ae58b805cbc0101fb2bc14eff9c379fc/lib/muack/mock.rb#L151-L154

``` ruby
if ::Class.instance_method(:method_defined?).arity == 1 # Ruby 2.5-
  def self.direct_method_defined? mod, msg
    mod.public_instance_methods(false).include?(msg) ||
      mod.protected_instance_methods(false).include?(msg) ||
      mod.private_instance_methods(false).include?(msg)
  end
else
  def self.direct_method_defined? mod, msg
    mod.method_defined?(msg, false) || # this doesn't cover private method
      mod.private_method_defined?(msg, false)
  end
end
```

For the third point, consider this case:

``` ruby
mod0 = Module.new{ def f; :m0; end }
mod1 = Module.new{ def f; :m1; end }
klass0 = Class.new{ prepend mod0 }
klass1 = Class.new{ prepend mod1 }
klass = Class.new{ prepend mod1; prepend mod0 }

allow_any_instance_of(klass).to receive(:f).and_return(:f)

expect(klass.new.f).to eq(:f)
expect(klass0.new.f).to eq(:m0)
expect(klass1.new.f).to eq(:m1)
```

File: lib/rspec/mocks/any_instance/recorder.rb [hunk 8]
```diff
@@ -257,9 +288,31 @@ def observe!(method_name)
...
+          if @prepended_module
+            unless @prepended_module.method_defined?(method_name)
+              original_method_name = build_alias_method_name(method_name)
+
+              # Defining the stubbed method as a proxy, calling further the stub specifically defined for the
+              # current class, or, if no such stub found, calling the original method
+              @prepended_module.__send__(:define_method, method_name) do |*args, &blk|
+                possible_method_stub = "__#{self.class.name}_#{method_name}_without_any_instance__"
+                if respond_to?(possible_method_stub)
+                  __send__(possible_method_stub, *args, &blk)
```
Comment (godfat) [inline lib/rspec/mocks/any_instance/recorder.rb]: Interesting approach! So far I can't really think of a downside for this, but I'll keep pondering if I can find a case where this cannot cover.

**Review by godfat — COMMENTED** [review]
Interesting approach. I left some comments about some implementation details, which would be nice to have some covering tests but I guess that is ok. I'll keep thinking on some tricky edge cases and see if this will work for all of them.

**Comment by godfat** [comment]
And sorry for the delay! I went on vacation and didn't really want to review this while I was taking the vacation 😅

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 8 above)
Comment (godfat) [reply lib/rspec/mocks/any_instance/recorder.rb]: Here's a test case for you:

``` ruby
mod = Module.new{ def f; :f; end }
base = Class.new{ prepend mod }
sub = Class.new(base)

allow_any_instance_of(base).to receive(:f).and_return(:g)

expect(base.new.f).to eq(:g)
expect(sub.new.f).to eq(:g)
```

This may show that `self.class` is not a sufficient key. The same goes to stubbing on `mod`, which all other instances might have all different classes yet since we're stubbing on `mod` it should affect all of them.

File: lib/rspec/mocks/any_instance/recorder.rb [hunk 9]
```diff
@@ -275,18 +328,9 @@ def mark_invoked!(method_name)
...
-            problem_mod = prepended_modules.find { |mod| mod.method_defined?(method_name) }
-            return unless problem_mod
-
-            AnyInstance.error_generator.raise_not_supported_with_prepend_error(method_name, problem_mod)
-          end
-        else
-          def allow_no_prepended_module_definition_of(_method_name)
-            # nothing to do; prepends aren't supported on this version of ruby
-          end
+        def prepended_module(method_name)
```
Comment (godfat) [inline lib/rspec/mocks/any_instance/recorder.rb:344]: My solution for this on Muack is this: https://github.com/godfat/muack/blob/e8bd6473ae58b805cbc0101fb2bc14eff9c379fc/lib/muack/mock.rb#L167-L190

``` ruby
def self.prepare_target singleton_class, msg
  if singleton_class == singleton_class.ancestors.first # no prepended mod
    singleton_class
  else # check if we need to prepend an internal module to override
    index = singleton_class.ancestors.index(singleton_class)
    prepended_modules = singleton_class.ancestors[0...index]

    if prepended_modules.find{ |m| direct_method_defined?(m, msg) }
      # prepend an internal module to override the prepended module(s)
      name = :MuackPrepended
      if singleton_class.const_defined?(name)
        singleton_class.const_get(name)
      else
        prepended = ::Module.new
        singleton_class.const_set(name, prepended)
        singleton_class.private_constant(name)
        singleton_class.prepend(prepended)
        prepended
      end
    else # we don't need to override so singleton class is enough
      singleton_class
    end
  end
end
```

I think this is not perfect because we'll be inserting an internal prepended module, but so far this solved any edge cases I can think of. I hope this can give you some inspiration.

**Comment by texpert** [comment]
Thank you for review and hints, @godfat ! Will try to further work on the PR as soon, as I'll get some time.

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 3 above)
Comment (texpert) [reply lib/rspec/mocks/any_instance/recorder.rb:130]: Cleaning implemented in https://github.com/godfat/rspec-mocks/pull/1/commits/70c2b88f6295361ed33a5a33959babc32d919a01

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 4 above)
Comment (texpert) [reply lib/rspec/mocks/any_instance/recorder.rb:132]: Nope, they're not the same. This is in the case of dynamically defined classes without a name - the object id's are changing, the class name remains the same. That's why I had to name somehow the class, if it has no name.

File: spec/rspec/mocks/any_instance_spec.rb (hunk 5 above)
Comment (texpert) [reply spec/rspec/mocks/any_instance_spec.rb]: Fixed in https://github.com/godfat/rspec-mocks/pull/1/commits/e30bf39612257c00b2f678f2d0f21a3cf4b032e4

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 6 above)
Comment (texpert) [reply lib/rspec/mocks/any_instance/recorder.rb:345]: Hm, but it IS based on method_name.

Could you, please, elaborate waht do you mean or how to test it?

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 7 above)
Comment (texpert) [reply lib/rspec/mocks/any_instance/recorder.rb]: Oops, as for why the `last` method - it was just a dumb semantic error - sure I've meant `first`, as we're dealing with a prepended module.

Fix in https://github.com/godfat/rspec-mocks/pull/1/commits/a352272418ce91d21e2ef1a8876001dfdd185168 and your case added to the specs.

Thanks a lot!!

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 8 above)
Comment (texpert) [reply lib/rspec/mocks/any_instance/recorder.rb]: I have fixed this too - https://github.com/godfat/rspec-mocks/pull/1/commits/db750748f0f5ebddbf9d53b83b907d741bf60268

**Comment by texpert** [comment]
I have still to work on restoring the original method visibility, but this will be another day when I'll got some time.

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 6 above)
Comment (godfat) [reply lib/rspec/mocks/any_instance/recorder.rb:345]: > Hm, but it IS based on method_name.

When you're accessing `@prepended_module`, it's not based on `method_name`, unless you're always using `prepended_module(method_name)` so you're always passing `method_name`, but then this memorization is always discarded and has no effect.

> Could you, please, elaborate waht do you mean or how to test it?

In practice, it can be possible that this disparity might not happen, which I don't know if it'll happen or not, but I think this is a code smell because the assumption is not consistent. When you access `@prepended_module`, you'll need to know which was the last time `prepended_module(method_name)` was called and if that `method_name` is matching the one you have in mind for this time. It's very easy to get wrong with this dependency.

File: lib/rspec/mocks/any_instance/recorder.rb [hunk 10]
```diff
@@ -275,18 +336,9 @@ def mark_invoked!(method_name)
...
+          @prepended_module =
+            RSpec::Mocks::Proxy.prepended_modules_of(@klass).select { |mod| mod.method_defined?(method_name) }.first
```
Comment (godfat) [inline lib/rspec/mocks/any_instance/recorder.rb]: I should probably mention this earlier, but only now I realized. For `select` and then `first`, that's `find`.

So this can be just:

``` ruby
RSpec::Mocks::Proxy.prepended_modules_of(@klass).find { |mod| mod.method_defined?(method_name) }
```

Comment (texpert) [reply lib/rspec/mocks/any_instance/recorder.rb]: Fixed in https://github.com/godfat/rspec-mocks/pull/1/commits/31768dd8659c23c2263ea2135e4998f879fed9bf

**Comment by texpert** [comment]
> I have still to work on restoring the original method visibility, but this will be another day when I'll got some time.

I have overlooked that everything's OK with preserving and restoring method's visibility.

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 6 above)
Comment (texpert) [reply lib/rspec/mocks/any_instance/recorder.rb:345]: OK, @godfat , I've added a test for stubbing methods defined in different prepended modules, and, also, added fetching the exact prepended_module where the method is defined in `restore_original_method!` method.

Is it good now?

Comment (godfat) [reply lib/rspec/mocks/any_instance/recorder.rb:345]: @texpert Sorry for not responding for so long. There were some bigger holidays and I was getting a bit more busy lately so I didn't look into this for quite some time. I'll get onto this again soon (in a few days maybe?), and definitely when the pull request you're basing off got merged upstream. It's sad that it takes so long and it's still not there yet, but I guess that's it. (as part of my excuse :s)

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 4 above)
Comment (godfat) [reply lib/rspec/mocks/any_instance/recorder.rb:132]: I should have responded to this earlier. Sorry for that.

I am confused why they are not the same, specifically given this line:

> the object id's are changing, the class name remains the same

Why is the id changing, yet its name remains the same? Here if `@klass.name` does not exist, meaning that it's a dynamically generated module, which we set the name based on `@klass.object_id`, and then we use `@klass.name`, doesn't that mean:

``` ruby
"__#{@klass.name}_#{method_name}_without_any_instance__"
```

is equivalent to:

``` ruby
"__Klass_#{@klass.object_id}_#{method_name}_without_any_instance__"
```

?

Could you please give an example when id would be changing, yet the name is the same? Do you mean when the user explicitly sets a name? In that case, I am not sure what we should really do. My thinking is, if the id is changing, doesn't mean they are different things? So they shouldn't share the same thing?

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 3 above)
Comment (godfat) [reply lib/rspec/mocks/any_instance/recorder.rb:130]: I'll prefer if this can be set inside a non-top level namespace, for example:

``` ruby
RSpec::Mocks.const_set("Klass_#{@klass.object_id}", @klass)
```

But this is less important.

File: lib/rspec/mocks/any_instance/recorder.rb [hunk 11]
```diff
@@ -196,31 +205,48 @@ def restore_original_method!(method_name)
           return unless @klass.instance_method(method_name).owner == @klass
 
           alias_method_name = build_alias_method_name(method_name)
-          @klass.class_exec(@backed_up_method_owner) do |backed_up_method_owner|
-            remove_method method_name
+          if prepended_module(method_name)
+            alias_prepended_method_name = build_alias_prepended_method_name(method_name)
+            module_methods = @prepended_module.public_instance_methods
+            remaining_stubs = module_methods.select { |m| m.end_with?("_#{method_name}_without_any_instance__") }
```
Comment (godfat) [inline lib/rspec/mocks/any_instance/recorder.rb:211]: A new idea here. First we can use `count` instead of `select` because it seems we don't really need those name, just how many there are.

Secondly, can we maybe count it internally, instead of observing it?

Every time we insert a new stub, we do `@remaining_stubs += 1` and every time we remove a stub, we do `@remaining_stubs -= 1`.

Generally, checking by counting names, which can easily hit into name conflicts, gives me some nerve.

File: lib/rspec/mocks/any_instance/recorder.rb [hunk 12]
```diff
@@ -196,31 +205,48 @@ def restore_original_method!(method_name)
...
+            @prepended_module.module_exec do
+              remove_method alias_prepended_method_name
+              Object.send(:remove_const, @klass.name) if @klass.name.start_with?('Klass_')
+
+              break if remaining_stubs.size > 1
```
Comment (godfat) [inline lib/rspec/mocks/any_instance/recorder.rb:216]: Could you please explain why we're always removing the module, but not always removing the method? (only when `remaining_stubs.size > 1`)

File: lib/rspec/mocks/any_instance/recorder.rb [hunk 13]
```diff
@@ -196,31 +205,48 @@ def restore_original_method!(method_name)
...
-            # For ruby 2.3 and above, we need to only restore methods that
-            # @klass originally owned.
-            #
-            # https://github.com/ruby/ruby/commit/c8854d2ca4be9ee6946e6d17b0e17d9ef130ee81
-            if RUBY_VERSION < "2.3" || backed_up_method_owner[method_name.to_sym] == self
+              remove_method method_name
               alias_method method_name, alias_method_name
+              remove_method alias_method_name
+            end
+            @prepended_module = nil if remaining_stubs.size == 1 # This was the last stub
```
Comment (godfat) [inline lib/rspec/mocks/any_instance/recorder.rb:222]: Shouldn't we remove the module only at this point?

File: lib/rspec/mocks/any_instance/recorder.rb [hunk 14]
```diff
@@ -235,8 +261,17 @@ def backup_method!(method_name)
 
           alias_method_name = build_alias_method_name(method_name)
           @backed_up_method_owner[method_name.to_sym] ||= @klass.instance_method(method_name).owner
-          @klass.class_exec do
-            alias_method alias_method_name, method_name
+          if prepended_module(method_name)
+            unless @prepended_module.method_defined?(alias_method_name)
+              @prepended_module.module_exec do
+                alias_method alias_method_name, method_name
+                remove_method method_name
```
Comment (godfat) [inline lib/rspec/mocks/any_instance/recorder.rb:268]: Why do we only remove method in this case (when there is `prepended_module`), but not for the other case?

File: lib/rspec/mocks/any_instance/recorder.rb (hunk 8 above)
Comment (godfat) [reply lib/rspec/mocks/any_instance/recorder.rb]: You covered my point 2, but didn't cover my point 0 and 1.

**Comment by godfat** [comment]
@texpert I am very sorry for the long delay, and I have less free time recently. Given that it's been awhile I am sure I must have forgotten some of the details and was probably missing a lot of context in my head. Apologies if some of my questions look a bit silly or out of places. 

To be honest, I am feeling a bit tired of reviewing patches against codes not maintained by me. I don't really understand how those work, thus it's difficult to review patches against them. In order to properly give a proper review, I have to understand the original codes first, and then I can continue to understand the patch.

I've done this in the past few rounds of reviews, which really took a lot of time. I think I'd done the right thing, but given that my pull request isn't even merged yet, I am not fully sure if the effort spent upfront is justified.

I think I will pause on very thorough review, and only give some quick read and review from now on. And I would really like to pass the review to the RSpec team, if possible.

I thank you for contributing patches based on my patch, I am honored by that. Let's hope https://github.com/rspec/rspec-mocks/pull/1218 can be merged soon and we can move on from there.

**Comment by godfat** [comment]
@texpert Now the base branch is merged into `4-0-dev` upstream, do you want to rebase and create a new pull request at https://github.com/rspec/rspec-mocks and also target `4-0-dev`, so that RSpec team members can take a look?

**Comment by texpert** [comment]
> @texpert Now the base branch is merged into `4-0-dev` upstream, do you want to rebase and create a new pull request at https://github.com/rspec/rspec-mocks and also target `4-0-dev`, so that RSpec team members can take a look?

Sure, @godfat - will find time either this, or next weekend.

Congratulations with the merge of your PR!

**Comment by godfat** [comment]
Thanks! I'll try to review your PR as well. Hopefully we can fix all the
wild corner cases.
